SSL enhancements including SNI support

dmlloyd opened the pull request:
Also various API updates to better support SSL/TLS use cases, and various cleanups on both the client and server side.

pskopek commented on src/main/java/org/wildfly/security/auth/login/SecurityDomain.java:
```diff
@@ -389,6 +417,20 @@ public Builder setPermissionMapper(PermissionMapper permissionMapper) {
+         * @return this builder
+         */
+        public Builder setPrincipalDecoder(PrincipalDecoder principalDecoder) {
+            Assert.checkNotNullParam("roleMapper", roleMapper);
```
This should be "principalDecoder", I guess.

dmlloyd replied:
Indeed...

dmlloyd commented:
Fixed that small problem.  Also, squashed the import fixes into the commits that introduced the problems, and squashed away the merge commits.  Should once again be possible to review commit-by-commit.
